src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/SpellCheck/LspCSharpSpellCheckRangeProvider.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.Hosting;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.SpellCheck;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.SpellCheck;
+
+internal sealed class LspCSharpSpellCheckRangeProvider(
+    LanguageServerFeatureOptions languageServerFeatureOptions,
+    IClientConnection clientConnection) : ICSharpSpellCheckRangeProvider
+{
+    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions = languageServerFeatureOptions;
+    private readonly IClientConnection _clientConnection = clientConnection;
+
+    public async Task<ImmutableArray<SpellCheckRange>> GetCSharpSpellCheckRangesAsync(DocumentContext documentContext, CancellationToken cancellationToken)
+    {
+        if (!_languageServerFeatureOptions.SingleServerSupport)
+        {
+            return [];
+        }
+
+        var delegatedParams = new DelegatedSpellCheckParams(documentContext.GetTextDocumentIdentifierAndVersion());
+        var delegatedResponse = await _clientConnection.SendRequestAsync<DelegatedSpellCheckParams, VSInternalSpellCheckableRangeReport[]?>(
+            CustomMessageNames.RazorSpellCheckEndpoint,
+            delegatedParams,
+            cancellationToken).ConfigureAwait(false);
+
+        if (delegatedResponse is null)
+        {
+            return [];
+        }
+
+        using var ranges = new PooledArrayBuilder<SpellCheckRange>();
+        foreach (var report in delegatedResponse)
+        {
+            if (report.Ranges is not { } csharpRanges)
+            {
+                continue;
+            }
+
+            // Since we get C# tokens that have relative starts, we need to convert them back to absolute indexes
+            // so we can sort them with the Razor tokens later
+            var absoluteCSharpStartIndex = 0;
+            for (var i = 0; i < csharpRanges.Length; i += 3)
+            {
+                var kind = csharpRanges[i];
+                var start = csharpRanges[i + 1];
+                var length = csharpRanges[i + 2];
+
+                absoluteCSharpStartIndex += start;
+
+                ranges.Add(new(kind, absoluteCSharpStartIndex, length));
+
+                absoluteCSharpStartIndex += length;
+            }
+        }
+
+        return ranges.ToImmutable();
```

Comment:
>ToImmutable [](http://example.com/codeflow?start=22&length=11)

Should this be DrainToImmutable? If I remember what Dustin said correctly possible perf benefits in some scenarios?

Reply:
`DrainToImmutable` has a perf advantage if the size of the array is known up front, but for the record I hate that we as developers have to think about that.

In this case, we don't know the size up front, but its probably a reasonable assumption to make that there will only be one report, so we can probably make a decent guess.